// traceur/dist/commonjs/util/CollectingErrorReporter.js
"use strict";
var $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__,
    $___46__46__47_util_47_ErrorReporter_46_js__;
var $__createClass = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ = require("traceur/dist/commonjs/runtime/modules/createClass.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__}).default;
var $__superConstructor = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ = require("traceur/dist/commonjs/runtime/modules/superConstructor.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__}).default;
var ErrorReporter = ($___46__46__47_util_47_ErrorReporter_46_js__ = require("../util/ErrorReporter.js"), $___46__46__47_util_47_ErrorReporter_46_js__ && $___46__46__47_util_47_ErrorReporter_46_js__.__esModule && $___46__46__47_util_47_ErrorReporter_46_js__ || {default: $___46__46__47_util_47_ErrorReporter_46_js__}).ErrorReporter;
var MultipleErrors = function($__super) {
  function MultipleErrors(errors) {
    $__superConstructor(MultipleErrors).call(this);
    this.message = errors ? errors.join('\n') : '';
    this.name = 'MultipleErrors';
    this.errors = errors;
  }
  return ($__createClass)(MultipleErrors, {}, {}, $__super);
}(Error);
var CollectingErrorReporter = function($__super) {
  function CollectingErrorReporter() {
    $__superConstructor(CollectingErrorReporter).call(this);
    this.errors = [];
  }
  return ($__createClass)(CollectingErrorReporter, {
    reportMessageInternal: function(location, message) {
      this.errors.push((location.start + ": " + message));
    },
    errorsAsString: function() {
      return this.toError().message;
    },
    toError: function() {
      return new MultipleErrors(this.errors);
    }
  }, {}, $__super);
}(ErrorReporter);
Object.defineProperties(module.exports, {
  MultipleErrors: {
    get: function() {
      return MultipleErrors;
    },
    enumerable: true
  },
  CollectingErrorReporter: {
    get: function() {
      return CollectingErrorReporter;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
