// traceur/dist/commonjs/util/StringSet.js
"use strict";
var $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__;
var $__createClass = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ = require("traceur/dist/commonjs/runtime/modules/createClass.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__}).default;
function assertString(value) {
  if (typeof value !== 'string')
    throw new TypeError();
}
var StringSet = function() {
  function StringSet() {
    this.storage_ = Object.create(null);
  }
  return ($__createClass)(StringSet, {
    add: function(value) {
      assertString(value);
      this.storage_[value] = true;
    },
    has: function(value) {
      assertString(value);
      return this.storage_[value] !== undefined;
    },
    delete: function(value) {
      assertString(value);
      delete this.storage_[value];
    },
    isEmpty: function() {
      for (var _ in this.storage_) {
        return false;
      }
      return true;
    },
    valuesAsArray: function() {
      return Object.keys(this.storage_);
    },
    forEach: function(func) {
      for (var value in this.storage_) {
        func(value);
      }
    }
  }, {});
}();
Object.defineProperties(module.exports, {
  StringSet: {
    get: function() {
      return StringSet;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
