// traceur/dist/commonjs/codegeneration/module/DirectExportVisitor.js
"use strict";
var $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__,
    $__ExportVisitor_46_js__;
var $__createClass = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ = require("traceur/dist/commonjs/runtime/modules/createClass.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__}).default;
var $__superConstructor = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ = require("traceur/dist/commonjs/runtime/modules/superConstructor.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__}).default;
var ExportVisitor = ($__ExportVisitor_46_js__ = require("./ExportVisitor.js"), $__ExportVisitor_46_js__ && $__ExportVisitor_46_js__.__esModule && $__ExportVisitor_46_js__ || {default: $__ExportVisitor_46_js__}).ExportVisitor;
var DirectExportVisitor = function($__super) {
  function DirectExportVisitor() {
    $__superConstructor(DirectExportVisitor).call(this, null, null, null);
    this.namedExports = [];
    this.starExports = [];
  }
  return ($__createClass)(DirectExportVisitor, {
    addExport: function(name, tree) {
      this.namedExports.push({
        name: name,
        tree: tree,
        moduleSpecifier: this.moduleSpecifier
      });
    },
    visitExportStar: function(tree) {
      this.starExports.push(this.moduleSpecifier);
    },
    hasExports: function() {
      return this.namedExports.length !== 0 || this.starExports.length !== 0;
    }
  }, {}, $__super);
}(ExportVisitor);
Object.defineProperties(module.exports, {
  DirectExportVisitor: {
    get: function() {
      return DirectExportVisitor;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
