// traceur/dist/commonjs/codegeneration/AsyncToGeneratorTransformer.js
"use strict";
var $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__,
    $___46__46__47_syntax_47_PredefinedName_46_js__,
    $__FindArguments_46_js__,
    $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__,
    $__ImportRuntimeTrait_46_js__,
    $__ParenTrait_46_js__,
    $__PlaceholderParser_46_js__,
    $__TempVarTransformer_46_js__,
    $__ParseTreeFactory_46_js__;
var $__createClass = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ = require("traceur/dist/commonjs/runtime/modules/createClass.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__}).default;
var $__superConstructor = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ = require("traceur/dist/commonjs/runtime/modules/superConstructor.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__}).default;
var $__superGet = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ = require("traceur/dist/commonjs/runtime/modules/superGet.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__}).default;
var $__getTemplateObject = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ = require("traceur/dist/commonjs/runtime/modules/getTemplateObject.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__}).default;
var ARGUMENTS = ($___46__46__47_syntax_47_PredefinedName_46_js__ = require("../syntax/PredefinedName.js"), $___46__46__47_syntax_47_PredefinedName_46_js__ && $___46__46__47_syntax_47_PredefinedName_46_js__.__esModule && $___46__46__47_syntax_47_PredefinedName_46_js__ || {default: $___46__46__47_syntax_47_PredefinedName_46_js__}).ARGUMENTS;
var FindArguments = ($__FindArguments_46_js__ = require("./FindArguments.js"), $__FindArguments_46_js__ && $__FindArguments_46_js__.__esModule && $__FindArguments_46_js__ || {default: $__FindArguments_46_js__}).default;
var $__7 = ($___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ = require("../syntax/trees/ParseTrees.js"), $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ && $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__.__esModule && $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ || {default: $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__}),
    FunctionBody = $__7.FunctionBody,
    FunctionDeclaration = $__7.FunctionDeclaration,
    FunctionExpression = $__7.FunctionExpression,
    Method = $__7.Method,
    YieldExpression = $__7.YieldExpression;
var ImportRuntimeTrait = ($__ImportRuntimeTrait_46_js__ = require("./ImportRuntimeTrait.js"), $__ImportRuntimeTrait_46_js__ && $__ImportRuntimeTrait_46_js__.__esModule && $__ImportRuntimeTrait_46_js__ || {default: $__ImportRuntimeTrait_46_js__}).default;
var ParenTrait = ($__ParenTrait_46_js__ = require("./ParenTrait.js"), $__ParenTrait_46_js__ && $__ParenTrait_46_js__.__esModule && $__ParenTrait_46_js__ || {default: $__ParenTrait_46_js__}).ParenTrait;
var parseStatement = ($__PlaceholderParser_46_js__ = require("./PlaceholderParser.js"), $__PlaceholderParser_46_js__ && $__PlaceholderParser_46_js__.__esModule && $__PlaceholderParser_46_js__ || {default: $__PlaceholderParser_46_js__}).parseStatement;
var TempVarTransformer = ($__TempVarTransformer_46_js__ = require("./TempVarTransformer.js"), $__TempVarTransformer_46_js__ && $__TempVarTransformer_46_js__.__esModule && $__TempVarTransformer_46_js__ || {default: $__TempVarTransformer_46_js__}).TempVarTransformer;
var $__12 = ($__ParseTreeFactory_46_js__ = require("./ParseTreeFactory.js"), $__ParseTreeFactory_46_js__ && $__ParseTreeFactory_46_js__.__esModule && $__ParseTreeFactory_46_js__ || {default: $__ParseTreeFactory_46_js__}),
    createIdentifierExpression = $__12.createIdentifierExpression,
    createNullLiteral = $__12.createNullLiteral;
var AsyncToGeneratorTransformer = function($__super) {
  function AsyncToGeneratorTransformer(identifierGenerator, reporter, options) {
    $__superConstructor(AsyncToGeneratorTransformer).call(this, identifierGenerator, reporter, options);
    this.inAsyncFunction_ = false;
  }
  return ($__createClass)(AsyncToGeneratorTransformer, {
    transformFunctionDeclaration: function(tree) {
      if (tree.isAsyncFunction()) {
        return this.transformFunctionShared_(tree, FunctionDeclaration);
      }
      return $__superGet(this, AsyncToGeneratorTransformer.prototype, "transformFunctionDeclaration").call(this, tree);
    },
    transformFunctionExpression: function(tree) {
      if (tree.isAsyncFunction()) {
        return this.transformFunctionShared_(tree, FunctionExpression);
      }
      return $__superGet(this, AsyncToGeneratorTransformer.prototype, "transformFunctionExpression").call(this, tree);
    },
    transformFunctionShared_: function(tree, ctor) {
      var parameterList = this.transformAny(tree.parameterList);
      var typeAnnotation = this.transformAny(tree.typeAnnotation);
      var annotations = this.transformList(tree.annotations);
      var body = this.transformAsyncBody_(tree.body);
      return new ctor(tree.location, tree.name, null, parameterList, typeAnnotation, annotations, body);
    },
    transformAsyncBody_: function(body) {
      var inAsyncFunction = this.inAsyncFunction_;
      this.inAsyncFunction_ = true;
      body = this.transformFunctionBody(body);
      var spawn = this.getRuntimeExpression('spawn');
      body = wrapBodyInSpawn(body, spawn);
      this.inAsyncFunction_ = inAsyncFunction;
      return body;
    },
    transformMethod: function(tree) {
      if (tree.isAsyncFunction()) {
        var name = this.transformAny(tree.name);
        var parameterList = this.transformAny(tree.parameterList);
        var typeAnnotation = this.transformAny(tree.typeAnnotation);
        var annotations = this.transformList(tree.annotations);
        var body = this.transformAsyncBody_(tree.body);
        return new Method(tree.location, tree.isStatic, null, name, parameterList, typeAnnotation, annotations, body, tree.debugName);
      }
      return $__superGet(this, AsyncToGeneratorTransformer.prototype, "transformMethod").call(this, tree);
    },
    transformAwaitExpression: function(tree) {
      if (this.inAsyncFunction_) {
        var expression = this.transformAny(tree.expression);
        return new YieldExpression(tree.location, expression, false);
      }
      return $__superGet(this, AsyncToGeneratorTransformer.prototype, "transformAwaitExpression").call(this, tree);
    }
  }, {}, $__super);
}(ImportRuntimeTrait(ParenTrait(TempVarTransformer)));
function wrapBodyInSpawn(body, spawn) {
  var visitor = new FindArguments();
  visitor.visitAny(body);
  var argExpr = visitor.found ? createIdentifierExpression(ARGUMENTS) : createNullLiteral();
  var statement = parseStatement($__getTemplateObject(["return ", "(this, ", ", function*() { ", " });"]), spawn, argExpr, body);
  return new FunctionBody(body.location, [statement]);
}
Object.defineProperties(module.exports, {
  AsyncToGeneratorTransformer: {
    get: function() {
      return AsyncToGeneratorTransformer;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
