// traceur/dist/commonjs/codegeneration/CommonJsModuleTransformer.js
"use strict";
var $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_spread_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__,
    $__ModuleTransformer_46_js__,
    $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__,
    $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__,
    $___46__46__47_util_47_assert_46_js__,
    $__globalThis_46_js__,
    $__PlaceholderParser_46_js__,
    $__ParseTreeFactory_46_js__,
    $___46__46__47_syntax_47_TokenType_46_js__;
var $__spread = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_spread_46_js__ = require("traceur/dist/commonjs/runtime/modules/spread.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_spread_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_spread_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_spread_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_spread_46_js__}).default;
var $__createClass = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ = require("traceur/dist/commonjs/runtime/modules/createClass.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__}).default;
var $__superConstructor = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ = require("traceur/dist/commonjs/runtime/modules/superConstructor.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__}).default;
var $__superGet = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ = require("traceur/dist/commonjs/runtime/modules/superGet.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__}).default;
var $__getTemplateObject = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ = require("traceur/dist/commonjs/runtime/modules/getTemplateObject.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__}).default;
var ModuleTransformer = ($__ModuleTransformer_46_js__ = require("./ModuleTransformer.js"), $__ModuleTransformer_46_js__ && $__ModuleTransformer_46_js__.__esModule && $__ModuleTransformer_46_js__ || {default: $__ModuleTransformer_46_js__}).ModuleTransformer;
var $__8 = ($___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__ = require("../syntax/trees/ParseTreeType.js"), $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__ && $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__.__esModule && $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__ || {default: $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__}),
    CALL_EXPRESSION = $__8.CALL_EXPRESSION,
    GET_ACCESSOR = $__8.GET_ACCESSOR,
    OBJECT_LITERAL = $__8.OBJECT_LITERAL,
    PROPERTY_NAME_ASSIGNMENT = $__8.PROPERTY_NAME_ASSIGNMENT,
    RETURN_STATEMENT = $__8.RETURN_STATEMENT;
var $__9 = ($___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ = require("../syntax/trees/ParseTrees.js"), $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ && $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__.__esModule && $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ || {default: $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__}),
    ArgumentList = $__9.ArgumentList,
    CallExpression = $__9.CallExpression,
    ExpressionStatement = $__9.ExpressionStatement;
var assert = ($___46__46__47_util_47_assert_46_js__ = require("../util/assert.js"), $___46__46__47_util_47_assert_46_js__ && $___46__46__47_util_47_assert_46_js__.__esModule && $___46__46__47_util_47_assert_46_js__ || {default: $___46__46__47_util_47_assert_46_js__}).assert;
var globalThis = ($__globalThis_46_js__ = require("./globalThis.js"), $__globalThis_46_js__ && $__globalThis_46_js__.__esModule && $__globalThis_46_js__ || {default: $__globalThis_46_js__}).default;
var $__12 = ($__PlaceholderParser_46_js__ = require("./PlaceholderParser.js"), $__PlaceholderParser_46_js__ && $__PlaceholderParser_46_js__.__esModule && $__PlaceholderParser_46_js__ || {default: $__PlaceholderParser_46_js__}),
    parseExpression = $__12.parseExpression,
    parsePropertyDefinition = $__12.parsePropertyDefinition,
    parseStatements = $__12.parseStatements;
var $__13 = ($__ParseTreeFactory_46_js__ = require("./ParseTreeFactory.js"), $__ParseTreeFactory_46_js__ && $__ParseTreeFactory_46_js__.__esModule && $__ParseTreeFactory_46_js__ || {default: $__ParseTreeFactory_46_js__}),
    createEmptyParameterList = $__13.createEmptyParameterList,
    createFunctionExpression = $__13.createFunctionExpression,
    createIdentifierExpression = $__13.createIdentifierExpression,
    createObjectLiteral = $__13.createObjectLiteral,
    createPropertyNameAssignment = $__13.createPropertyNameAssignment,
    createVariableStatement = $__13.createVariableStatement,
    createVariableDeclaration = $__13.createVariableDeclaration,
    createVariableDeclarationList = $__13.createVariableDeclarationList;
var VAR = ($___46__46__47_syntax_47_TokenType_46_js__ = require("../syntax/TokenType.js"), $___46__46__47_syntax_47_TokenType_46_js__ && $___46__46__47_syntax_47_TokenType_46_js__.__esModule && $___46__46__47_syntax_47_TokenType_46_js__ || {default: $___46__46__47_syntax_47_TokenType_46_js__}).VAR;
var CommonJsModuleTransformer = function($__super) {
  function CommonJsModuleTransformer(identifierGenerator, reporter) {
    var options = arguments[2];
    $__superConstructor(CommonJsModuleTransformer).call(this, identifierGenerator, reporter, options);
    this.moduleVars_ = [];
    this.anonymousModule = options && !options.bundle && options.moduleName !== true;
  }
  return ($__createClass)(CommonJsModuleTransformer, {
    getModuleName: function(tree) {
      if (this.anonymousModule)
        return null;
      return tree.moduleName;
    },
    moduleProlog: function() {
      var statements = $__superGet(this, CommonJsModuleTransformer.prototype, "moduleProlog").call(this);
      if (this.moduleVars_.length) {
        var tmpVarDeclarations = createVariableStatement(createVariableDeclarationList(VAR, this.moduleVars_.map(function(varName) {
          return createVariableDeclaration(varName, null);
        })));
        statements.push(tmpVarDeclarations);
      }
      return statements;
    },
    wrapModule: function(statements) {
      var last = statements[statements.length - 1];
      statements = statements.slice(0, -1);
      assert(last.type === RETURN_STATEMENT);
      var exportExpression = last.expression;
      if (this.hasExports()) {
        var exportStatement = this.transformExportExpressionToModuleExport(exportExpression);
        statements = statements.concat(exportStatement);
      }
      return statements;
    },
    transformExportExpressionToModuleExport: function(tree) {
      var expression;
      if (tree.type === CALL_EXPRESSION) {
        var descriptors = this.transformObjectLiteralToDescriptors(tree.args.args[0]);
        var object = parseExpression($__getTemplateObject(["Object.defineProperties(module.exports, ", ")"]), descriptors);
        var newArgs = new ArgumentList(tree.args.location, $__spread([object], tree.args.args.slice(1)));
        expression = new CallExpression(tree.location, tree.operand, newArgs);
      } else {
        var descriptors$__1 = this.transformObjectLiteralToDescriptors(tree);
        expression = parseExpression($__getTemplateObject(["Object.defineProperties(module.exports, ", ")"]), descriptors$__1);
      }
      return new ExpressionStatement(expression.location, expression);
    },
    transformObjectLiteralToDescriptors: function(literalTree) {
      assert(literalTree.type === OBJECT_LITERAL);
      var props = literalTree.propertyNameAndValues.map(function(exp) {
        var descriptor;
        switch (exp.type) {
          case GET_ACCESSOR:
            {
              var getterFunction = createFunctionExpression(createEmptyParameterList(), exp.body);
              descriptor = parseExpression($__getTemplateObject(["{get: ", ", enumerable: true}"]), getterFunction);
              break;
            }
          case PROPERTY_NAME_ASSIGNMENT:
            descriptor = parseExpression($__getTemplateObject(["{value: ", "}"]), exp.value);
            break;
          default:
            throw new Error(("Unexpected property type " + exp.type));
        }
        return createPropertyNameAssignment(exp.name, descriptor);
      });
      return createObjectLiteral(props);
    },
    transformModuleSpecifier: function(tree) {
      var moduleName = tree.token.processedValue;
      var tmpVar = this.getTempVarNameForModuleSpecifier(tree);
      this.moduleVars_.push(tmpVar);
      var tvId = createIdentifierExpression(tmpVar);
      return parseExpression($__getTemplateObject(["(", " = require(", "),\n        ", " && ", ".__esModule && ", " || {default: ", "})"]), tvId, moduleName, tvId, tvId, tvId, tvId);
    },
    getExportProperties: function() {
      var properties = $__superGet(this, CommonJsModuleTransformer.prototype, "getExportProperties").call(this);
      if (this.exportVisitor_.hasExports())
        properties.push(parsePropertyDefinition($__getTemplateObject(["__esModule: true"])));
      return properties;
    }
  }, {}, $__super);
}(ModuleTransformer);
Object.defineProperties(module.exports, {
  CommonJsModuleTransformer: {
    get: function() {
      return CommonJsModuleTransformer;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
