// traceur/dist/commonjs/codegeneration/GeneratorTransformPass.js
"use strict";
var $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__,
    $__ArrowFunctionTransformer_46_js__,
    $__generator_47_AsyncTransformer_46_js__,
    $__generator_47_ForInTransformPass_46_js__,
    $__generator_47_GeneratorTransformer_46_js__,
    $__PlaceholderParser_46_js__,
    $__TempVarTransformer_46_js__,
    $__ImportRuntimeTrait_46_js__,
    $__FindInFunctionScope_46_js__,
    $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__,
    $__ParseTreeFactory_46_js__;
var $__createClass = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ = require("traceur/dist/commonjs/runtime/modules/createClass.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__}).default;
var $__superConstructor = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ = require("traceur/dist/commonjs/runtime/modules/superConstructor.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__}).default;
var $__superGet = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ = require("traceur/dist/commonjs/runtime/modules/superGet.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__}).default;
var $__getTemplateObject = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ = require("traceur/dist/commonjs/runtime/modules/getTemplateObject.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__}).default;
var ArrowFunctionTransformer = ($__ArrowFunctionTransformer_46_js__ = require("./ArrowFunctionTransformer.js"), $__ArrowFunctionTransformer_46_js__ && $__ArrowFunctionTransformer_46_js__.__esModule && $__ArrowFunctionTransformer_46_js__ || {default: $__ArrowFunctionTransformer_46_js__}).ArrowFunctionTransformer;
var AsyncTransformer = ($__generator_47_AsyncTransformer_46_js__ = require("./generator/AsyncTransformer.js"), $__generator_47_AsyncTransformer_46_js__ && $__generator_47_AsyncTransformer_46_js__.__esModule && $__generator_47_AsyncTransformer_46_js__ || {default: $__generator_47_AsyncTransformer_46_js__}).AsyncTransformer;
var ForInTransformPass = ($__generator_47_ForInTransformPass_46_js__ = require("./generator/ForInTransformPass.js"), $__generator_47_ForInTransformPass_46_js__ && $__generator_47_ForInTransformPass_46_js__.__esModule && $__generator_47_ForInTransformPass_46_js__ || {default: $__generator_47_ForInTransformPass_46_js__}).ForInTransformPass;
var GeneratorTransformer = ($__generator_47_GeneratorTransformer_46_js__ = require("./generator/GeneratorTransformer.js"), $__generator_47_GeneratorTransformer_46_js__ && $__generator_47_GeneratorTransformer_46_js__.__esModule && $__generator_47_GeneratorTransformer_46_js__ || {default: $__generator_47_GeneratorTransformer_46_js__}).GeneratorTransformer;
var $__12 = ($__PlaceholderParser_46_js__ = require("./PlaceholderParser.js"), $__PlaceholderParser_46_js__ && $__PlaceholderParser_46_js__.__esModule && $__PlaceholderParser_46_js__ || {default: $__PlaceholderParser_46_js__}),
    parseExpression = $__12.parseExpression,
    parseStatement = $__12.parseStatement;
var TempVarTransformer = ($__TempVarTransformer_46_js__ = require("./TempVarTransformer.js"), $__TempVarTransformer_46_js__ && $__TempVarTransformer_46_js__.__esModule && $__TempVarTransformer_46_js__ || {default: $__TempVarTransformer_46_js__}).TempVarTransformer;
var ImportRuntimeTrait = ($__ImportRuntimeTrait_46_js__ = require("./ImportRuntimeTrait.js"), $__ImportRuntimeTrait_46_js__ && $__ImportRuntimeTrait_46_js__.__esModule && $__ImportRuntimeTrait_46_js__ || {default: $__ImportRuntimeTrait_46_js__}).default;
var FindInFunctionScope = ($__FindInFunctionScope_46_js__ = require("./FindInFunctionScope.js"), $__FindInFunctionScope_46_js__ && $__FindInFunctionScope_46_js__.__esModule && $__FindInFunctionScope_46_js__ || {default: $__FindInFunctionScope_46_js__}).FindInFunctionScope;
var $__16 = ($___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ = require("../syntax/trees/ParseTrees.js"), $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ && $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__.__esModule && $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ || {default: $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__}),
    AnonBlock = $__16.AnonBlock,
    FunctionDeclaration = $__16.FunctionDeclaration,
    FunctionExpression = $__16.FunctionExpression;
var $__17 = ($__ParseTreeFactory_46_js__ = require("./ParseTreeFactory.js"), $__ParseTreeFactory_46_js__ && $__ParseTreeFactory_46_js__.__esModule && $__ParseTreeFactory_46_js__ || {default: $__ParseTreeFactory_46_js__}),
    createBindingIdentifier = $__17.createBindingIdentifier,
    id = $__17.createIdentifierExpression,
    createIdentifierToken = $__17.createIdentifierToken;
var ForInFinder = function($__super) {
  function ForInFinder() {
    $__superConstructor(ForInFinder).apply(this, arguments);
  }
  return ($__createClass)(ForInFinder, {visitForInStatement: function(tree) {
      this.found = true;
    }}, {}, $__super);
}(FindInFunctionScope);
function needsTransform(tree, transformOptions) {
  return transformOptions.generators && tree.isGenerator() || transformOptions.asyncFunctions && tree.isAsyncFunction();
}
var GeneratorTransformPass = function($__super) {
  function GeneratorTransformPass(identifierGenerator, reporter, options) {
    $__superConstructor(GeneratorTransformPass).call(this, identifierGenerator, reporter, options);
    this.tranformOptions_ = options.transformOptions;
    this.inBlock_ = false;
  }
  return ($__createClass)(GeneratorTransformPass, {
    transformFunctionDeclaration: function(tree) {
      if (!needsTransform(tree, this.tranformOptions_))
        return $__superGet(this, GeneratorTransformPass.prototype, "transformFunctionDeclaration").call(this, tree);
      if (tree.isGenerator())
        return this.transformGeneratorDeclaration_(tree);
      return this.transformFunction_(tree, FunctionDeclaration, null);
    },
    transformGeneratorDeclaration_: function(tree) {
      var nameIdExpression = id(tree.name.identifierToken);
      var initGeneratorFunction = this.getRuntimeExpression('initGeneratorFunction');
      var setupPrototypeExpression = parseExpression($__getTemplateObject(["", "(", ")"]), initGeneratorFunction, nameIdExpression);
      var tmpVar = id(this.inBlock_ ? this.getTempIdentifier() : this.addTempVar(setupPrototypeExpression));
      var funcDecl = this.transformFunction_(tree, FunctionDeclaration, tmpVar);
      if (!this.inBlock_)
        return funcDecl;
      return new AnonBlock(null, [funcDecl, parseStatement($__getTemplateObject(["var ", " = ", ""]), tmpVar, setupPrototypeExpression)]);
    },
    transformFunctionExpression: function(tree) {
      if (!needsTransform(tree, this.tranformOptions_))
        return $__superGet(this, GeneratorTransformPass.prototype, "transformFunctionExpression").call(this, tree);
      if (tree.isGenerator())
        return this.transformGeneratorExpression_(tree);
      return this.transformFunction_(tree, FunctionExpression, null);
    },
    transformGeneratorExpression_: function(tree) {
      var name;
      if (!tree.name) {
        name = createIdentifierToken(this.getTempIdentifier());
        tree = new FunctionExpression(tree.location, createBindingIdentifier(name), tree.functionKind, tree.parameterList, tree.typeAnnotation, tree.annotations, tree.body);
      } else {
        name = tree.name.identifierToken;
      }
      var functionExpression = this.transformFunction_(tree, FunctionExpression, id(name));
      var initGeneratorFunction = this.getRuntimeExpression('initGeneratorFunction');
      return parseExpression($__getTemplateObject(["", "(", ")"]), initGeneratorFunction, functionExpression);
    },
    transformFunction_: function(tree, constructor, nameExpression) {
      var $__2 = this;
      var body = $__superGet(this, GeneratorTransformPass.prototype, "transformAny").call(this, tree.body);
      var finder = new ForInFinder();
      finder.visitAny(body);
      if (finder.found) {
        body = new ForInTransformPass(this.identifierGenerator, this.reporter, this.options).transformAny(body);
      }
      if (this.tranformOptions_.generators && tree.isGenerator()) {
        var transformer = new GeneratorTransformer(this.identifierGenerator, this.reporter, this.options);
        body = transformer.transformGeneratorBody(body, nameExpression);
        transformer.requiredNames.forEach(function(n) {
          $__2.addImportedName(n);
        });
      } else if (this.tranformOptions_.asyncFunctions && tree.isAsyncFunction()) {
        var transformer$__3 = new AsyncTransformer(this.identifierGenerator, this.reporter, this.options);
        body = transformer$__3.transformAsyncBody(body, nameExpression);
        transformer$__3.requiredNames.forEach(function(n) {
          $__2.addImportedName(n);
        });
      }
      var functionKind = null;
      return new constructor(tree.location, tree.name, functionKind, tree.parameterList, tree.typeAnnotation || null, tree.annotations || null, body);
    },
    transformArrowFunction: function(tree) {
      if (!tree.isAsyncFunction())
        return $__superGet(this, GeneratorTransformPass.prototype, "transformArrowFunction").call(this, tree);
      return this.transformAny(ArrowFunctionTransformer.transform(this, tree));
    },
    transformBlock: function(tree) {
      var inBlock = this.inBlock_;
      this.inBlock_ = true;
      var rv = $__superGet(this, GeneratorTransformPass.prototype, "transformBlock").call(this, tree);
      this.inBlock_ = inBlock;
      return rv;
    }
  }, {}, $__super);
}(ImportRuntimeTrait(TempVarTransformer));
Object.defineProperties(module.exports, {
  GeneratorTransformPass: {
    get: function() {
      return GeneratorTransformPass;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
