// traceur/dist/commonjs/codegeneration/MemberVariableConstructorTransformer.js
"use strict";
var $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__,
    $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__,
    $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__,
    $__ParseTreeTransformer_46_js__,
    $__ParseTreeFactory_46_js__,
    $__PrependStatements_46_js__;
var $__createClass = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ = require("traceur/dist/commonjs/runtime/modules/createClass.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__}).default;
var $__superConstructor = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ = require("traceur/dist/commonjs/runtime/modules/superConstructor.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__}).default;
var $__superGet = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ = require("traceur/dist/commonjs/runtime/modules/superGet.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__}).default;
var Method = ($___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ = require("../syntax/trees/ParseTrees.js"), $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ && $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__.__esModule && $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ || {default: $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__}).Method;
var SUPER_EXPRESSION = ($___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__ = require("../syntax/trees/ParseTreeType.js"), $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__ && $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__.__esModule && $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__ || {default: $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__}).SUPER_EXPRESSION;
var ParseTreeTransformer = ($__ParseTreeTransformer_46_js__ = require("./ParseTreeTransformer.js"), $__ParseTreeTransformer_46_js__ && $__ParseTreeTransformer_46_js__.__esModule && $__ParseTreeTransformer_46_js__ || {default: $__ParseTreeTransformer_46_js__}).ParseTreeTransformer;
var $__7 = ($__ParseTreeFactory_46_js__ = require("./ParseTreeFactory.js"), $__ParseTreeFactory_46_js__ && $__ParseTreeFactory_46_js__.__esModule && $__ParseTreeFactory_46_js__ || {default: $__ParseTreeFactory_46_js__}),
    createCommaExpression = $__7.createCommaExpression,
    createExpressionStatement = $__7.createExpressionStatement,
    createFunctionBody = $__7.createFunctionBody,
    createParenExpression = $__7.createParenExpression,
    createThisExpression = $__7.createThisExpression;
var prependStatements = ($__PrependStatements_46_js__ = require("./PrependStatements.js"), $__PrependStatements_46_js__ && $__PrependStatements_46_js__.__esModule && $__PrependStatements_46_js__ || {default: $__PrependStatements_46_js__}).prependStatements;
function transformConstructor(constructor, initExpression, superClass) {
  if (superClass) {
    var transformer = new SuperCallTransformer(initExpression);
    return transformer.transformAny(constructor);
  }
  var statements = constructor.body.statements;
  var initStatement = createExpressionStatement(initExpression);
  statements = prependStatements(statements, initStatement);
  return new Method(constructor.location, false, constructor.functionKind, constructor.name, constructor.parameterList, constructor.typeAnnotation, constructor.annotations, createFunctionBody(statements), constructor.debugName);
}
var SuperCallTransformer = function($__super) {
  function SuperCallTransformer(expression) {
    $__superConstructor(SuperCallTransformer).call(this);
    this.expression = expression;
  }
  return ($__createClass)(SuperCallTransformer, {
    transformCallExpression: function(tree) {
      if (tree.operand.type === SUPER_EXPRESSION) {
        var thisExpression = createThisExpression();
        return createParenExpression(createCommaExpression([tree, this.expression, thisExpression]));
      }
      return $__superGet(this, SuperCallTransformer.prototype, "transformCallExpression").call(this, tree);
    },
    transformClassDeclaration: function(tree) {
      return tree;
    },
    transformClassExpression: function(tree) {
      return tree;
    }
  }, {}, $__super);
}(ParseTreeTransformer);
Object.defineProperties(module.exports, {
  transformConstructor: {
    get: function() {
      return transformConstructor;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
