// traceur/dist/commonjs/codegeneration/MultiTransformer.js
"use strict";
var $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__,
    $___46__46__47_syntax_47_ParseTreeValidator_46_js__;
var $__createClass = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ = require("traceur/dist/commonjs/runtime/modules/createClass.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__}).default;
var ParseTreeValidator = ($___46__46__47_syntax_47_ParseTreeValidator_46_js__ = require("../syntax/ParseTreeValidator.js"), $___46__46__47_syntax_47_ParseTreeValidator_46_js__ && $___46__46__47_syntax_47_ParseTreeValidator_46_js__.__esModule && $___46__46__47_syntax_47_ParseTreeValidator_46_js__ || {default: $___46__46__47_syntax_47_ParseTreeValidator_46_js__}).ParseTreeValidator;
var MultiTransformer = function() {
  function MultiTransformer(reporter, validate) {
    this.reporter_ = reporter;
    this.validate_ = validate;
    this.treeTransformers_ = [];
  }
  return ($__createClass)(MultiTransformer, {
    append: function(treeTransformer) {
      this.treeTransformers_.push(treeTransformer);
    },
    transform: function(tree) {
      var reporter = this.reporter_;
      var validate = this.validate_;
      this.treeTransformers_.every(function(transformTree) {
        tree = transformTree(tree);
        if (reporter.hadError())
          return false;
        if (validate)
          ParseTreeValidator.validate(tree);
        return true;
      });
      return tree;
    }
  }, {});
}();
Object.defineProperties(module.exports, {
  MultiTransformer: {
    get: function() {
      return MultiTransformer;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
