// traceur/dist/commonjs/codegeneration/ProperTailCallTransformer.js
"use strict";
var $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__,
    $__TempVarTransformer_46_js__,
    $__RewriteTailCallsTransformer_46_js__,
    $__ParseTreeFactory_46_js__,
    $__PlaceholderParser_46_js__,
    $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__,
    $__ImportRuntimeTrait_46_js__;
var $__createClass = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ = require("traceur/dist/commonjs/runtime/modules/createClass.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__}).default;
var $__superConstructor = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ = require("traceur/dist/commonjs/runtime/modules/superConstructor.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__}).default;
var $__superGet = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ = require("traceur/dist/commonjs/runtime/modules/superGet.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__}).default;
var $__getTemplateObject = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ = require("traceur/dist/commonjs/runtime/modules/getTemplateObject.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_getTemplateObject_46_js__}).default;
var TempVarTransformer = ($__TempVarTransformer_46_js__ = require("./TempVarTransformer.js"), $__TempVarTransformer_46_js__ && $__TempVarTransformer_46_js__.__esModule && $__TempVarTransformer_46_js__ || {default: $__TempVarTransformer_46_js__}).TempVarTransformer;
var RewriteTailCallsTransformer = ($__RewriteTailCallsTransformer_46_js__ = require("./RewriteTailCallsTransformer.js"), $__RewriteTailCallsTransformer_46_js__ && $__RewriteTailCallsTransformer_46_js__.__esModule && $__RewriteTailCallsTransformer_46_js__ || {default: $__RewriteTailCallsTransformer_46_js__}).RewriteTailCallsTransformer;
var $__7 = ($__ParseTreeFactory_46_js__ = require("./ParseTreeFactory.js"), $__ParseTreeFactory_46_js__ && $__ParseTreeFactory_46_js__.__esModule && $__ParseTreeFactory_46_js__ || {default: $__ParseTreeFactory_46_js__}),
    createFunctionBody = $__7.createFunctionBody,
    createFunctionExpression = $__7.createFunctionExpression,
    id = $__7.createIdentifierExpression;
var $__8 = ($__PlaceholderParser_46_js__ = require("./PlaceholderParser.js"), $__PlaceholderParser_46_js__ && $__PlaceholderParser_46_js__.__esModule && $__PlaceholderParser_46_js__ || {default: $__PlaceholderParser_46_js__}),
    parseExpression = $__8.parseExpression,
    parseStatement = $__8.parseStatement,
    parseStatements = $__8.parseStatements;
var $__9 = ($___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ = require("../syntax/trees/ParseTrees.js"), $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ && $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__.__esModule && $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__ || {default: $___46__46__47_syntax_47_trees_47_ParseTrees_46_js__}),
    AnonBlock = $__9.AnonBlock,
    FunctionDeclaration = $__9.FunctionDeclaration,
    FunctionExpression = $__9.FunctionExpression;
var ImportRuntimeTrait = ($__ImportRuntimeTrait_46_js__ = require("./ImportRuntimeTrait.js"), $__ImportRuntimeTrait_46_js__ && $__ImportRuntimeTrait_46_js__.__esModule && $__ImportRuntimeTrait_46_js__ || {default: $__ImportRuntimeTrait_46_js__}).default;
var ProperTailCallTransformer = function($__super) {
  function ProperTailCallTransformer(identifierGenerator, reporter, options) {
    $__superConstructor(ProperTailCallTransformer).call(this, identifierGenerator, reporter, options);
    this.inBlock_ = false;
    this.options = options;
  }
  return ($__createClass)(ProperTailCallTransformer, {
    transformFunctionDeclaration: function(tree) {
      tree = $__superGet(this, ProperTailCallTransformer.prototype, "transformFunctionDeclaration").call(this, tree);
      if (tree.functionKind !== null) {
        return tree;
      }
      var nameIdExpression = id(tree.name.identifierToken);
      var initTailRecursiveFunction = this.getRuntimeExpression('initTailRecursiveFunction');
      var setupFlagExpression = parseExpression($__getTemplateObject(["", "(", ")"]), initTailRecursiveFunction, nameIdExpression);
      var funcDecl = this.transformFunction_(tree, FunctionDeclaration);
      if (funcDecl === tree) {
        return tree;
      }
      var tmpVar = id(this.inBlock_ ? this.getTempIdentifier() : this.addTempVar(setupFlagExpression));
      if (!this.inBlock_) {
        return funcDecl;
      }
      return new AnonBlock(null, [funcDecl, parseStatement($__getTemplateObject(["var ", " = ", ";"]), tmpVar, setupFlagExpression)]);
    },
    transformFunctionExpression: function(tree) {
      tree = $__superGet(this, ProperTailCallTransformer.prototype, "transformFunctionExpression").call(this, tree);
      if (tree.functionKind) {
        return tree;
      }
      var functionExpression = this.transformFunction_(tree, FunctionExpression);
      if (functionExpression === tree) {
        return tree;
      }
      var initTailRecursiveFunction = this.getRuntimeExpression('initTailRecursiveFunction');
      return parseExpression($__getTemplateObject(["", "(", ")"]), initTailRecursiveFunction, functionExpression);
    },
    transformFunction_: function(tree, constructor) {
      var body = RewriteTailCallsTransformer.transform(this, tree.body);
      if (body === tree.body) {
        return tree;
      }
      var func = id(this.getTempIdentifier());
      var innerFunction = createFunctionExpression(tree.parameterList, body);
      var call = this.getRuntimeExpression('call');
      var outerBody = createFunctionBody(parseStatements($__getTemplateObject(["\n        return ", "(", ", this, arguments);"]), call, innerFunction));
      return new constructor(tree.location, tree.name, tree.functionKind, tree.parameterList, tree.typeAnnotation, tree.annotations, outerBody);
    },
    transformBlock: function(tree) {
      var inBlock = this.inBlock_;
      this.inBlock_ = true;
      var rv = $__superGet(this, ProperTailCallTransformer.prototype, "transformBlock").call(this, tree);
      this.inBlock_ = inBlock;
      return rv;
    }
  }, {}, $__super);
}(ImportRuntimeTrait(TempVarTransformer));
Object.defineProperties(module.exports, {
  ProperTailCallTransformer: {
    get: function() {
      return ProperTailCallTransformer;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
