// traceur/dist/commonjs/semantics/ConstructorValidator.js
"use strict";
var $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__,
    $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__,
    $___46__46__47_codegeneration_47_FindVisitor_46_js__;
var $__createClass = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ = require("traceur/dist/commonjs/runtime/modules/createClass.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__}).default;
var $__superConstructor = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ = require("traceur/dist/commonjs/runtime/modules/superConstructor.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__}).default;
var $__superGet = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ = require("traceur/dist/commonjs/runtime/modules/superGet.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superGet_46_js__}).default;
var SUPER_EXPRESSION = ($___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__ = require("../syntax/trees/ParseTreeType.js"), $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__ && $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__.__esModule && $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__ || {default: $___46__46__47_syntax_47_trees_47_ParseTreeType_46_js__}).SUPER_EXPRESSION;
var FindVisitor = ($___46__46__47_codegeneration_47_FindVisitor_46_js__ = require("../codegeneration/FindVisitor.js"), $___46__46__47_codegeneration_47_FindVisitor_46_js__ && $___46__46__47_codegeneration_47_FindVisitor_46_js__.__esModule && $___46__46__47_codegeneration_47_FindVisitor_46_js__ || {default: $___46__46__47_codegeneration_47_FindVisitor_46_js__}).FindVisitor;
var ConstructorValidator = function($__super) {
  function ConstructorValidator(reporter) {
    $__superConstructor(ConstructorValidator).call(this);
    this.reporter_ = reporter;
    this.hasError = false;
  }
  return ($__createClass)(ConstructorValidator, {
    visitClassExpression: function(tree) {
      this.visitAny(tree.superClass);
    },
    visitClassDeclaration: function(tree) {
      this.visitAny(tree.superClass);
    },
    visitThisExpression: function(tree) {
      this.reportError_(tree.location, 'this');
    },
    visitCallExpression: function(tree) {
      if (tree.operand.type === SUPER_EXPRESSION) {
        this.visitAny(tree.args);
        this.found = true;
        return;
      }
      $__superGet(this, ConstructorValidator.prototype, "visitCallExpression").call(this, tree);
    },
    visitSuperExpression: function(tree) {
      this.reportError_(tree.location, 'super property');
    },
    reportError_: function(location, kind) {
      this.reporter_.reportError(location, ("'" + kind + "' is not allowed before super()"));
      this.hasError = true;
      this.found = true;
    }
  }, {}, $__super);
}(FindVisitor);
function validateConstructor(tree, reporter) {
  var visitor = new ConstructorValidator(reporter);
  visitor.visitAny(tree);
  if (visitor.hasError)
    return false;
  if (visitor.found)
    return true;
  reporter.reportError(tree.location, 'Derived constructor must call super()');
  return false;
}
Object.defineProperties(module.exports, {
  validateConstructor: {
    get: function() {
      return validateConstructor;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
