// traceur/dist/commonjs/runtime/private.js
"use strict";
var $__private_45_symbol_46_js__,
    $__private_45_weak_45_map_46_js__;
var sym = ($__private_45_symbol_46_js__ = require("./private-symbol.js"), $__private_45_symbol_46_js__ && $__private_45_symbol_46_js__.__esModule && $__private_45_symbol_46_js__ || {default: $__private_45_symbol_46_js__});
var weak = ($__private_45_weak_45_map_46_js__ = require("./private-weak-map.js"), $__private_45_weak_45_map_46_js__ && $__private_45_weak_45_map_46_js__.__esModule && $__private_45_weak_45_map_46_js__ || {default: $__private_45_weak_45_map_46_js__});
var hasWeakMap = typeof WeakMap === 'function';
var m = hasWeakMap ? weak : sym;
var isPrivateSymbol = m.isPrivateSymbol;
var createPrivateSymbol = m.createPrivateSymbol;
var hasPrivate = m.hasPrivate;
var deletePrivate = m.deletePrivate;
var setPrivate = m.setPrivate;
var getPrivate = m.getPrivate;
m.init();
Object.defineProperties(module.exports, {
  isPrivateSymbol: {
    get: function() {
      return isPrivateSymbol;
    },
    enumerable: true
  },
  createPrivateSymbol: {
    get: function() {
      return createPrivateSymbol;
    },
    enumerable: true
  },
  hasPrivate: {
    get: function() {
      return hasPrivate;
    },
    enumerable: true
  },
  deletePrivate: {
    get: function() {
      return deletePrivate;
    },
    enumerable: true
  },
  setPrivate: {
    get: function() {
      return setPrivate;
    },
    enumerable: true
  },
  getPrivate: {
    get: function() {
      return getPrivate;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
