// traceur/dist/commonjs/runtime/private-weak-map.js
"use strict";
var $WeakMap = typeof WeakMap === 'function' ? WeakMap : undefined;
function isPrivateSymbol(s) {
  return false;
}
function createPrivateSymbol() {
  return new $WeakMap();
}
function hasPrivate(obj, sym) {
  return sym.has(obj);
}
function deletePrivate(obj, sym) {
  return sym.delete(obj);
}
function setPrivate(obj, sym, val) {
  sym.set(obj, val);
}
function getPrivate(obj, sym) {
  return sym.get(obj);
}
function init() {}
Object.defineProperties(module.exports, {
  isPrivateSymbol: {
    get: function() {
      return isPrivateSymbol;
    },
    enumerable: true
  },
  createPrivateSymbol: {
    get: function() {
      return createPrivateSymbol;
    },
    enumerable: true
  },
  hasPrivate: {
    get: function() {
      return hasPrivate;
    },
    enumerable: true
  },
  deletePrivate: {
    get: function() {
      return deletePrivate;
    },
    enumerable: true
  },
  setPrivate: {
    get: function() {
      return setPrivate;
    },
    enumerable: true
  },
  getPrivate: {
    get: function() {
      return getPrivate;
    },
    enumerable: true
  },
  init: {
    get: function() {
      return init;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
