// traceur/dist/commonjs/syntax/PredefinedName.js
"use strict";
var $ARGUMENTS = '$arguments';
var ANY = 'any';
var APPLY = 'apply';
var ARGUMENTS = 'arguments';
var ARRAY = 'Array';
var AS = 'as';
var ASYNC = 'async';
var ASYNC_STAR = 'async star';
var AWAIT = 'await';
var BIND = 'bind';
var CALL = 'call';
var CONFIGURABLE = 'configurable';
var CONSTRUCTOR = 'constructor';
var CREATE = 'create';
var CURRENT = 'current';
var DEFINE_PROPERTY = 'defineProperty';
var ENUMERABLE = 'enumerable';
var FREEZE = 'freeze';
var FROM = 'from';
var FUNCTION = 'Function';
var GET = 'get';
var HAS = 'has';
var LENGTH = 'length';
var MODULE = 'module';
var NEW = 'new';
var OBJECT = 'Object';
var OBJECT_NAME = 'Object';
var OF = 'of';
var ON = 'on';
var PREVENT_EXTENSIONS = 'preventExtensions';
var PROTOTYPE = 'prototype';
var PUSH = 'push';
var SET = 'set';
var SLICE = 'slice';
var THIS = 'this';
var TRACEUR_RUNTIME = '$traceurRuntime';
var TYPE = 'type';
var UNDEFINED = 'undefined';
var WRITABLE = 'writable';
Object.defineProperties(module.exports, {
  $ARGUMENTS: {
    get: function() {
      return $ARGUMENTS;
    },
    enumerable: true
  },
  ANY: {
    get: function() {
      return ANY;
    },
    enumerable: true
  },
  APPLY: {
    get: function() {
      return APPLY;
    },
    enumerable: true
  },
  ARGUMENTS: {
    get: function() {
      return ARGUMENTS;
    },
    enumerable: true
  },
  ARRAY: {
    get: function() {
      return ARRAY;
    },
    enumerable: true
  },
  AS: {
    get: function() {
      return AS;
    },
    enumerable: true
  },
  ASYNC: {
    get: function() {
      return ASYNC;
    },
    enumerable: true
  },
  ASYNC_STAR: {
    get: function() {
      return ASYNC_STAR;
    },
    enumerable: true
  },
  AWAIT: {
    get: function() {
      return AWAIT;
    },
    enumerable: true
  },
  BIND: {
    get: function() {
      return BIND;
    },
    enumerable: true
  },
  CALL: {
    get: function() {
      return CALL;
    },
    enumerable: true
  },
  CONFIGURABLE: {
    get: function() {
      return CONFIGURABLE;
    },
    enumerable: true
  },
  CONSTRUCTOR: {
    get: function() {
      return CONSTRUCTOR;
    },
    enumerable: true
  },
  CREATE: {
    get: function() {
      return CREATE;
    },
    enumerable: true
  },
  CURRENT: {
    get: function() {
      return CURRENT;
    },
    enumerable: true
  },
  DEFINE_PROPERTY: {
    get: function() {
      return DEFINE_PROPERTY;
    },
    enumerable: true
  },
  ENUMERABLE: {
    get: function() {
      return ENUMERABLE;
    },
    enumerable: true
  },
  FREEZE: {
    get: function() {
      return FREEZE;
    },
    enumerable: true
  },
  FROM: {
    get: function() {
      return FROM;
    },
    enumerable: true
  },
  FUNCTION: {
    get: function() {
      return FUNCTION;
    },
    enumerable: true
  },
  GET: {
    get: function() {
      return GET;
    },
    enumerable: true
  },
  HAS: {
    get: function() {
      return HAS;
    },
    enumerable: true
  },
  LENGTH: {
    get: function() {
      return LENGTH;
    },
    enumerable: true
  },
  MODULE: {
    get: function() {
      return MODULE;
    },
    enumerable: true
  },
  NEW: {
    get: function() {
      return NEW;
    },
    enumerable: true
  },
  OBJECT: {
    get: function() {
      return OBJECT;
    },
    enumerable: true
  },
  OBJECT_NAME: {
    get: function() {
      return OBJECT_NAME;
    },
    enumerable: true
  },
  OF: {
    get: function() {
      return OF;
    },
    enumerable: true
  },
  ON: {
    get: function() {
      return ON;
    },
    enumerable: true
  },
  PREVENT_EXTENSIONS: {
    get: function() {
      return PREVENT_EXTENSIONS;
    },
    enumerable: true
  },
  PROTOTYPE: {
    get: function() {
      return PROTOTYPE;
    },
    enumerable: true
  },
  PUSH: {
    get: function() {
      return PUSH;
    },
    enumerable: true
  },
  SET: {
    get: function() {
      return SET;
    },
    enumerable: true
  },
  SLICE: {
    get: function() {
      return SLICE;
    },
    enumerable: true
  },
  THIS: {
    get: function() {
      return THIS;
    },
    enumerable: true
  },
  TRACEUR_RUNTIME: {
    get: function() {
      return TRACEUR_RUNTIME;
    },
    enumerable: true
  },
  TYPE: {
    get: function() {
      return TYPE;
    },
    enumerable: true
  },
  UNDEFINED: {
    get: function() {
      return UNDEFINED;
    },
    enumerable: true
  },
  WRITABLE: {
    get: function() {
      return WRITABLE;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
