// traceur/dist/commonjs/syntax/trees/ParseTreeType.js
"use strict";
var ANNOTATION = 'ANNOTATION';
var ANON_BLOCK = 'ANON_BLOCK';
var ARGUMENT_LIST = 'ARGUMENT_LIST';
var ARRAY_COMPREHENSION = 'ARRAY_COMPREHENSION';
var ARRAY_LITERAL = 'ARRAY_LITERAL';
var ARRAY_PATTERN = 'ARRAY_PATTERN';
var ARRAY_TYPE = 'ARRAY_TYPE';
var ARROW_FUNCTION = 'ARROW_FUNCTION';
var ASSIGNMENT_ELEMENT = 'ASSIGNMENT_ELEMENT';
var AWAIT_EXPRESSION = 'AWAIT_EXPRESSION';
var BINARY_EXPRESSION = 'BINARY_EXPRESSION';
var BINDING_ELEMENT = 'BINDING_ELEMENT';
var BINDING_IDENTIFIER = 'BINDING_IDENTIFIER';
var BLOCK = 'BLOCK';
var BREAK_STATEMENT = 'BREAK_STATEMENT';
var CALL_EXPRESSION = 'CALL_EXPRESSION';
var CALL_SIGNATURE = 'CALL_SIGNATURE';
var CASE_CLAUSE = 'CASE_CLAUSE';
var CATCH = 'CATCH';
var CLASS_DECLARATION = 'CLASS_DECLARATION';
var CLASS_EXPRESSION = 'CLASS_EXPRESSION';
var COMMA_EXPRESSION = 'COMMA_EXPRESSION';
var COMPREHENSION_FOR = 'COMPREHENSION_FOR';
var COMPREHENSION_IF = 'COMPREHENSION_IF';
var COMPUTED_PROPERTY_NAME = 'COMPUTED_PROPERTY_NAME';
var CONDITIONAL_EXPRESSION = 'CONDITIONAL_EXPRESSION';
var CONSTRUCT_SIGNATURE = 'CONSTRUCT_SIGNATURE';
var CONSTRUCTOR_TYPE = 'CONSTRUCTOR_TYPE';
var CONTINUE_STATEMENT = 'CONTINUE_STATEMENT';
var COVER_FORMALS = 'COVER_FORMALS';
var COVER_INITIALIZED_NAME = 'COVER_INITIALIZED_NAME';
var DEBUGGER_STATEMENT = 'DEBUGGER_STATEMENT';
var DEFAULT_CLAUSE = 'DEFAULT_CLAUSE';
var DO_WHILE_STATEMENT = 'DO_WHILE_STATEMENT';
var EMPTY_STATEMENT = 'EMPTY_STATEMENT';
var EXPORT_DECLARATION = 'EXPORT_DECLARATION';
var EXPORT_DEFAULT = 'EXPORT_DEFAULT';
var EXPORT_SPECIFIER = 'EXPORT_SPECIFIER';
var EXPORT_SPECIFIER_SET = 'EXPORT_SPECIFIER_SET';
var EXPORT_STAR = 'EXPORT_STAR';
var EXPRESSION_STATEMENT = 'EXPRESSION_STATEMENT';
var FINALLY = 'FINALLY';
var FOR_IN_STATEMENT = 'FOR_IN_STATEMENT';
var FOR_OF_STATEMENT = 'FOR_OF_STATEMENT';
var FOR_ON_STATEMENT = 'FOR_ON_STATEMENT';
var FOR_STATEMENT = 'FOR_STATEMENT';
var FORMAL_PARAMETER = 'FORMAL_PARAMETER';
var FORMAL_PARAMETER_LIST = 'FORMAL_PARAMETER_LIST';
var FORWARD_DEFAULT_EXPORT = 'FORWARD_DEFAULT_EXPORT';
var FUNCTION_BODY = 'FUNCTION_BODY';
var FUNCTION_DECLARATION = 'FUNCTION_DECLARATION';
var FUNCTION_EXPRESSION = 'FUNCTION_EXPRESSION';
var FUNCTION_TYPE = 'FUNCTION_TYPE';
var GENERATOR_COMPREHENSION = 'GENERATOR_COMPREHENSION';
var GET_ACCESSOR = 'GET_ACCESSOR';
var IDENTIFIER_EXPRESSION = 'IDENTIFIER_EXPRESSION';
var IF_STATEMENT = 'IF_STATEMENT';
var IMPORT_CLAUSE_PAIR = 'IMPORT_CLAUSE_PAIR';
var IMPORT_DECLARATION = 'IMPORT_DECLARATION';
var IMPORT_SPECIFIER = 'IMPORT_SPECIFIER';
var IMPORT_SPECIFIER_SET = 'IMPORT_SPECIFIER_SET';
var IMPORT_TYPE_CLAUSE = 'IMPORT_TYPE_CLAUSE';
var IMPORTED_BINDING = 'IMPORTED_BINDING';
var INDEX_SIGNATURE = 'INDEX_SIGNATURE';
var INTERFACE_DECLARATION = 'INTERFACE_DECLARATION';
var JSX_ATTRIBUTE = 'JSX_ATTRIBUTE';
var JSX_ELEMENT = 'JSX_ELEMENT';
var JSX_ELEMENT_NAME = 'JSX_ELEMENT_NAME';
var JSX_PLACEHOLDER = 'JSX_PLACEHOLDER';
var JSX_SPREAD_ATTRIBUTE = 'JSX_SPREAD_ATTRIBUTE';
var JSX_TEXT = 'JSX_TEXT';
var LABELLED_STATEMENT = 'LABELLED_STATEMENT';
var LITERAL_EXPRESSION = 'LITERAL_EXPRESSION';
var LITERAL_PROPERTY_NAME = 'LITERAL_PROPERTY_NAME';
var MEMBER_EXPRESSION = 'MEMBER_EXPRESSION';
var MEMBER_LOOKUP_EXPRESSION = 'MEMBER_LOOKUP_EXPRESSION';
var METHOD = 'METHOD';
var METHOD_SIGNATURE = 'METHOD_SIGNATURE';
var MODULE = 'MODULE';
var MODULE_SPECIFIER = 'MODULE_SPECIFIER';
var NAME_SPACE_EXPORT = 'NAME_SPACE_EXPORT';
var NAME_SPACE_IMPORT = 'NAME_SPACE_IMPORT';
var NAMED_EXPORT = 'NAMED_EXPORT';
var NEW_EXPRESSION = 'NEW_EXPRESSION';
var OBJECT_LITERAL = 'OBJECT_LITERAL';
var OBJECT_PATTERN = 'OBJECT_PATTERN';
var OBJECT_PATTERN_FIELD = 'OBJECT_PATTERN_FIELD';
var OBJECT_TYPE = 'OBJECT_TYPE';
var PAREN_EXPRESSION = 'PAREN_EXPRESSION';
var POSTFIX_EXPRESSION = 'POSTFIX_EXPRESSION';
var PREDEFINED_TYPE = 'PREDEFINED_TYPE';
var PROPERTY_NAME_ASSIGNMENT = 'PROPERTY_NAME_ASSIGNMENT';
var PROPERTY_NAME_SHORTHAND = 'PROPERTY_NAME_SHORTHAND';
var PROPERTY_SIGNATURE = 'PROPERTY_SIGNATURE';
var PROPERTY_VARIABLE_DECLARATION = 'PROPERTY_VARIABLE_DECLARATION';
var REST_PARAMETER = 'REST_PARAMETER';
var RETURN_STATEMENT = 'RETURN_STATEMENT';
var SCRIPT = 'SCRIPT';
var SET_ACCESSOR = 'SET_ACCESSOR';
var SPREAD_EXPRESSION = 'SPREAD_EXPRESSION';
var SPREAD_PATTERN_ELEMENT = 'SPREAD_PATTERN_ELEMENT';
var STATE_MACHINE = 'STATE_MACHINE';
var SUPER_EXPRESSION = 'SUPER_EXPRESSION';
var SWITCH_STATEMENT = 'SWITCH_STATEMENT';
var SYNTAX_ERROR_TREE = 'SYNTAX_ERROR_TREE';
var TEMPLATE_LITERAL_EXPRESSION = 'TEMPLATE_LITERAL_EXPRESSION';
var TEMPLATE_LITERAL_PORTION = 'TEMPLATE_LITERAL_PORTION';
var TEMPLATE_SUBSTITUTION = 'TEMPLATE_SUBSTITUTION';
var THIS_EXPRESSION = 'THIS_EXPRESSION';
var THROW_STATEMENT = 'THROW_STATEMENT';
var TRY_STATEMENT = 'TRY_STATEMENT';
var TYPE_ALIAS_DECLARATION = 'TYPE_ALIAS_DECLARATION';
var TYPE_ARGUMENTS = 'TYPE_ARGUMENTS';
var TYPE_NAME = 'TYPE_NAME';
var TYPE_PARAMETER = 'TYPE_PARAMETER';
var TYPE_PARAMETERS = 'TYPE_PARAMETERS';
var TYPE_REFERENCE = 'TYPE_REFERENCE';
var UNARY_EXPRESSION = 'UNARY_EXPRESSION';
var UNION_TYPE = 'UNION_TYPE';
var VARIABLE_DECLARATION = 'VARIABLE_DECLARATION';
var VARIABLE_DECLARATION_LIST = 'VARIABLE_DECLARATION_LIST';
var VARIABLE_STATEMENT = 'VARIABLE_STATEMENT';
var WHILE_STATEMENT = 'WHILE_STATEMENT';
var WITH_STATEMENT = 'WITH_STATEMENT';
var YIELD_EXPRESSION = 'YIELD_EXPRESSION';
Object.defineProperties(module.exports, {
  ANNOTATION: {
    get: function() {
      return ANNOTATION;
    },
    enumerable: true
  },
  ANON_BLOCK: {
    get: function() {
      return ANON_BLOCK;
    },
    enumerable: true
  },
  ARGUMENT_LIST: {
    get: function() {
      return ARGUMENT_LIST;
    },
    enumerable: true
  },
  ARRAY_COMPREHENSION: {
    get: function() {
      return ARRAY_COMPREHENSION;
    },
    enumerable: true
  },
  ARRAY_LITERAL: {
    get: function() {
      return ARRAY_LITERAL;
    },
    enumerable: true
  },
  ARRAY_PATTERN: {
    get: function() {
      return ARRAY_PATTERN;
    },
    enumerable: true
  },
  ARRAY_TYPE: {
    get: function() {
      return ARRAY_TYPE;
    },
    enumerable: true
  },
  ARROW_FUNCTION: {
    get: function() {
      return ARROW_FUNCTION;
    },
    enumerable: true
  },
  ASSIGNMENT_ELEMENT: {
    get: function() {
      return ASSIGNMENT_ELEMENT;
    },
    enumerable: true
  },
  AWAIT_EXPRESSION: {
    get: function() {
      return AWAIT_EXPRESSION;
    },
    enumerable: true
  },
  BINARY_EXPRESSION: {
    get: function() {
      return BINARY_EXPRESSION;
    },
    enumerable: true
  },
  BINDING_ELEMENT: {
    get: function() {
      return BINDING_ELEMENT;
    },
    enumerable: true
  },
  BINDING_IDENTIFIER: {
    get: function() {
      return BINDING_IDENTIFIER;
    },
    enumerable: true
  },
  BLOCK: {
    get: function() {
      return BLOCK;
    },
    enumerable: true
  },
  BREAK_STATEMENT: {
    get: function() {
      return BREAK_STATEMENT;
    },
    enumerable: true
  },
  CALL_EXPRESSION: {
    get: function() {
      return CALL_EXPRESSION;
    },
    enumerable: true
  },
  CALL_SIGNATURE: {
    get: function() {
      return CALL_SIGNATURE;
    },
    enumerable: true
  },
  CASE_CLAUSE: {
    get: function() {
      return CASE_CLAUSE;
    },
    enumerable: true
  },
  CATCH: {
    get: function() {
      return CATCH;
    },
    enumerable: true
  },
  CLASS_DECLARATION: {
    get: function() {
      return CLASS_DECLARATION;
    },
    enumerable: true
  },
  CLASS_EXPRESSION: {
    get: function() {
      return CLASS_EXPRESSION;
    },
    enumerable: true
  },
  COMMA_EXPRESSION: {
    get: function() {
      return COMMA_EXPRESSION;
    },
    enumerable: true
  },
  COMPREHENSION_FOR: {
    get: function() {
      return COMPREHENSION_FOR;
    },
    enumerable: true
  },
  COMPREHENSION_IF: {
    get: function() {
      return COMPREHENSION_IF;
    },
    enumerable: true
  },
  COMPUTED_PROPERTY_NAME: {
    get: function() {
      return COMPUTED_PROPERTY_NAME;
    },
    enumerable: true
  },
  CONDITIONAL_EXPRESSION: {
    get: function() {
      return CONDITIONAL_EXPRESSION;
    },
    enumerable: true
  },
  CONSTRUCT_SIGNATURE: {
    get: function() {
      return CONSTRUCT_SIGNATURE;
    },
    enumerable: true
  },
  CONSTRUCTOR_TYPE: {
    get: function() {
      return CONSTRUCTOR_TYPE;
    },
    enumerable: true
  },
  CONTINUE_STATEMENT: {
    get: function() {
      return CONTINUE_STATEMENT;
    },
    enumerable: true
  },
  COVER_FORMALS: {
    get: function() {
      return COVER_FORMALS;
    },
    enumerable: true
  },
  COVER_INITIALIZED_NAME: {
    get: function() {
      return COVER_INITIALIZED_NAME;
    },
    enumerable: true
  },
  DEBUGGER_STATEMENT: {
    get: function() {
      return DEBUGGER_STATEMENT;
    },
    enumerable: true
  },
  DEFAULT_CLAUSE: {
    get: function() {
      return DEFAULT_CLAUSE;
    },
    enumerable: true
  },
  DO_WHILE_STATEMENT: {
    get: function() {
      return DO_WHILE_STATEMENT;
    },
    enumerable: true
  },
  EMPTY_STATEMENT: {
    get: function() {
      return EMPTY_STATEMENT;
    },
    enumerable: true
  },
  EXPORT_DECLARATION: {
    get: function() {
      return EXPORT_DECLARATION;
    },
    enumerable: true
  },
  EXPORT_DEFAULT: {
    get: function() {
      return EXPORT_DEFAULT;
    },
    enumerable: true
  },
  EXPORT_SPECIFIER: {
    get: function() {
      return EXPORT_SPECIFIER;
    },
    enumerable: true
  },
  EXPORT_SPECIFIER_SET: {
    get: function() {
      return EXPORT_SPECIFIER_SET;
    },
    enumerable: true
  },
  EXPORT_STAR: {
    get: function() {
      return EXPORT_STAR;
    },
    enumerable: true
  },
  EXPRESSION_STATEMENT: {
    get: function() {
      return EXPRESSION_STATEMENT;
    },
    enumerable: true
  },
  FINALLY: {
    get: function() {
      return FINALLY;
    },
    enumerable: true
  },
  FOR_IN_STATEMENT: {
    get: function() {
      return FOR_IN_STATEMENT;
    },
    enumerable: true
  },
  FOR_OF_STATEMENT: {
    get: function() {
      return FOR_OF_STATEMENT;
    },
    enumerable: true
  },
  FOR_ON_STATEMENT: {
    get: function() {
      return FOR_ON_STATEMENT;
    },
    enumerable: true
  },
  FOR_STATEMENT: {
    get: function() {
      return FOR_STATEMENT;
    },
    enumerable: true
  },
  FORMAL_PARAMETER: {
    get: function() {
      return FORMAL_PARAMETER;
    },
    enumerable: true
  },
  FORMAL_PARAMETER_LIST: {
    get: function() {
      return FORMAL_PARAMETER_LIST;
    },
    enumerable: true
  },
  FORWARD_DEFAULT_EXPORT: {
    get: function() {
      return FORWARD_DEFAULT_EXPORT;
    },
    enumerable: true
  },
  FUNCTION_BODY: {
    get: function() {
      return FUNCTION_BODY;
    },
    enumerable: true
  },
  FUNCTION_DECLARATION: {
    get: function() {
      return FUNCTION_DECLARATION;
    },
    enumerable: true
  },
  FUNCTION_EXPRESSION: {
    get: function() {
      return FUNCTION_EXPRESSION;
    },
    enumerable: true
  },
  FUNCTION_TYPE: {
    get: function() {
      return FUNCTION_TYPE;
    },
    enumerable: true
  },
  GENERATOR_COMPREHENSION: {
    get: function() {
      return GENERATOR_COMPREHENSION;
    },
    enumerable: true
  },
  GET_ACCESSOR: {
    get: function() {
      return GET_ACCESSOR;
    },
    enumerable: true
  },
  IDENTIFIER_EXPRESSION: {
    get: function() {
      return IDENTIFIER_EXPRESSION;
    },
    enumerable: true
  },
  IF_STATEMENT: {
    get: function() {
      return IF_STATEMENT;
    },
    enumerable: true
  },
  IMPORT_CLAUSE_PAIR: {
    get: function() {
      return IMPORT_CLAUSE_PAIR;
    },
    enumerable: true
  },
  IMPORT_DECLARATION: {
    get: function() {
      return IMPORT_DECLARATION;
    },
    enumerable: true
  },
  IMPORT_SPECIFIER: {
    get: function() {
      return IMPORT_SPECIFIER;
    },
    enumerable: true
  },
  IMPORT_SPECIFIER_SET: {
    get: function() {
      return IMPORT_SPECIFIER_SET;
    },
    enumerable: true
  },
  IMPORT_TYPE_CLAUSE: {
    get: function() {
      return IMPORT_TYPE_CLAUSE;
    },
    enumerable: true
  },
  IMPORTED_BINDING: {
    get: function() {
      return IMPORTED_BINDING;
    },
    enumerable: true
  },
  INDEX_SIGNATURE: {
    get: function() {
      return INDEX_SIGNATURE;
    },
    enumerable: true
  },
  INTERFACE_DECLARATION: {
    get: function() {
      return INTERFACE_DECLARATION;
    },
    enumerable: true
  },
  JSX_ATTRIBUTE: {
    get: function() {
      return JSX_ATTRIBUTE;
    },
    enumerable: true
  },
  JSX_ELEMENT: {
    get: function() {
      return JSX_ELEMENT;
    },
    enumerable: true
  },
  JSX_ELEMENT_NAME: {
    get: function() {
      return JSX_ELEMENT_NAME;
    },
    enumerable: true
  },
  JSX_PLACEHOLDER: {
    get: function() {
      return JSX_PLACEHOLDER;
    },
    enumerable: true
  },
  JSX_SPREAD_ATTRIBUTE: {
    get: function() {
      return JSX_SPREAD_ATTRIBUTE;
    },
    enumerable: true
  },
  JSX_TEXT: {
    get: function() {
      return JSX_TEXT;
    },
    enumerable: true
  },
  LABELLED_STATEMENT: {
    get: function() {
      return LABELLED_STATEMENT;
    },
    enumerable: true
  },
  LITERAL_EXPRESSION: {
    get: function() {
      return LITERAL_EXPRESSION;
    },
    enumerable: true
  },
  LITERAL_PROPERTY_NAME: {
    get: function() {
      return LITERAL_PROPERTY_NAME;
    },
    enumerable: true
  },
  MEMBER_EXPRESSION: {
    get: function() {
      return MEMBER_EXPRESSION;
    },
    enumerable: true
  },
  MEMBER_LOOKUP_EXPRESSION: {
    get: function() {
      return MEMBER_LOOKUP_EXPRESSION;
    },
    enumerable: true
  },
  METHOD: {
    get: function() {
      return METHOD;
    },
    enumerable: true
  },
  METHOD_SIGNATURE: {
    get: function() {
      return METHOD_SIGNATURE;
    },
    enumerable: true
  },
  MODULE: {
    get: function() {
      return MODULE;
    },
    enumerable: true
  },
  MODULE_SPECIFIER: {
    get: function() {
      return MODULE_SPECIFIER;
    },
    enumerable: true
  },
  NAME_SPACE_EXPORT: {
    get: function() {
      return NAME_SPACE_EXPORT;
    },
    enumerable: true
  },
  NAME_SPACE_IMPORT: {
    get: function() {
      return NAME_SPACE_IMPORT;
    },
    enumerable: true
  },
  NAMED_EXPORT: {
    get: function() {
      return NAMED_EXPORT;
    },
    enumerable: true
  },
  NEW_EXPRESSION: {
    get: function() {
      return NEW_EXPRESSION;
    },
    enumerable: true
  },
  OBJECT_LITERAL: {
    get: function() {
      return OBJECT_LITERAL;
    },
    enumerable: true
  },
  OBJECT_PATTERN: {
    get: function() {
      return OBJECT_PATTERN;
    },
    enumerable: true
  },
  OBJECT_PATTERN_FIELD: {
    get: function() {
      return OBJECT_PATTERN_FIELD;
    },
    enumerable: true
  },
  OBJECT_TYPE: {
    get: function() {
      return OBJECT_TYPE;
    },
    enumerable: true
  },
  PAREN_EXPRESSION: {
    get: function() {
      return PAREN_EXPRESSION;
    },
    enumerable: true
  },
  POSTFIX_EXPRESSION: {
    get: function() {
      return POSTFIX_EXPRESSION;
    },
    enumerable: true
  },
  PREDEFINED_TYPE: {
    get: function() {
      return PREDEFINED_TYPE;
    },
    enumerable: true
  },
  PROPERTY_NAME_ASSIGNMENT: {
    get: function() {
      return PROPERTY_NAME_ASSIGNMENT;
    },
    enumerable: true
  },
  PROPERTY_NAME_SHORTHAND: {
    get: function() {
      return PROPERTY_NAME_SHORTHAND;
    },
    enumerable: true
  },
  PROPERTY_SIGNATURE: {
    get: function() {
      return PROPERTY_SIGNATURE;
    },
    enumerable: true
  },
  PROPERTY_VARIABLE_DECLARATION: {
    get: function() {
      return PROPERTY_VARIABLE_DECLARATION;
    },
    enumerable: true
  },
  REST_PARAMETER: {
    get: function() {
      return REST_PARAMETER;
    },
    enumerable: true
  },
  RETURN_STATEMENT: {
    get: function() {
      return RETURN_STATEMENT;
    },
    enumerable: true
  },
  SCRIPT: {
    get: function() {
      return SCRIPT;
    },
    enumerable: true
  },
  SET_ACCESSOR: {
    get: function() {
      return SET_ACCESSOR;
    },
    enumerable: true
  },
  SPREAD_EXPRESSION: {
    get: function() {
      return SPREAD_EXPRESSION;
    },
    enumerable: true
  },
  SPREAD_PATTERN_ELEMENT: {
    get: function() {
      return SPREAD_PATTERN_ELEMENT;
    },
    enumerable: true
  },
  STATE_MACHINE: {
    get: function() {
      return STATE_MACHINE;
    },
    enumerable: true
  },
  SUPER_EXPRESSION: {
    get: function() {
      return SUPER_EXPRESSION;
    },
    enumerable: true
  },
  SWITCH_STATEMENT: {
    get: function() {
      return SWITCH_STATEMENT;
    },
    enumerable: true
  },
  SYNTAX_ERROR_TREE: {
    get: function() {
      return SYNTAX_ERROR_TREE;
    },
    enumerable: true
  },
  TEMPLATE_LITERAL_EXPRESSION: {
    get: function() {
      return TEMPLATE_LITERAL_EXPRESSION;
    },
    enumerable: true
  },
  TEMPLATE_LITERAL_PORTION: {
    get: function() {
      return TEMPLATE_LITERAL_PORTION;
    },
    enumerable: true
  },
  TEMPLATE_SUBSTITUTION: {
    get: function() {
      return TEMPLATE_SUBSTITUTION;
    },
    enumerable: true
  },
  THIS_EXPRESSION: {
    get: function() {
      return THIS_EXPRESSION;
    },
    enumerable: true
  },
  THROW_STATEMENT: {
    get: function() {
      return THROW_STATEMENT;
    },
    enumerable: true
  },
  TRY_STATEMENT: {
    get: function() {
      return TRY_STATEMENT;
    },
    enumerable: true
  },
  TYPE_ALIAS_DECLARATION: {
    get: function() {
      return TYPE_ALIAS_DECLARATION;
    },
    enumerable: true
  },
  TYPE_ARGUMENTS: {
    get: function() {
      return TYPE_ARGUMENTS;
    },
    enumerable: true
  },
  TYPE_NAME: {
    get: function() {
      return TYPE_NAME;
    },
    enumerable: true
  },
  TYPE_PARAMETER: {
    get: function() {
      return TYPE_PARAMETER;
    },
    enumerable: true
  },
  TYPE_PARAMETERS: {
    get: function() {
      return TYPE_PARAMETERS;
    },
    enumerable: true
  },
  TYPE_REFERENCE: {
    get: function() {
      return TYPE_REFERENCE;
    },
    enumerable: true
  },
  UNARY_EXPRESSION: {
    get: function() {
      return UNARY_EXPRESSION;
    },
    enumerable: true
  },
  UNION_TYPE: {
    get: function() {
      return UNION_TYPE;
    },
    enumerable: true
  },
  VARIABLE_DECLARATION: {
    get: function() {
      return VARIABLE_DECLARATION;
    },
    enumerable: true
  },
  VARIABLE_DECLARATION_LIST: {
    get: function() {
      return VARIABLE_DECLARATION_LIST;
    },
    enumerable: true
  },
  VARIABLE_STATEMENT: {
    get: function() {
      return VARIABLE_STATEMENT;
    },
    enumerable: true
  },
  WHILE_STATEMENT: {
    get: function() {
      return WHILE_STATEMENT;
    },
    enumerable: true
  },
  WITH_STATEMENT: {
    get: function() {
      return WITH_STATEMENT;
    },
    enumerable: true
  },
  YIELD_EXPRESSION: {
    get: function() {
      return YIELD_EXPRESSION;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
