// traceur/dist/commonjs/node/NodeLoaderCompiler.js
"use strict";
var $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__,
    $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__,
    $___46__46__47_loader_47_LoaderCompiler_46_js__;
var $__createClass = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ = require("traceur/dist/commonjs/runtime/modules/createClass.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_createClass_46_js__}).default;
var $__superConstructor = ($__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ = require("traceur/dist/commonjs/runtime/modules/superConstructor.js"), $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__.__esModule && $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__ || {default: $__traceur_47_dist_47_commonjs_47_runtime_47_modules_47_superConstructor_46_js__}).default;
var LoaderCompiler = ($___46__46__47_loader_47_LoaderCompiler_46_js__ = require("../loader/LoaderCompiler.js"), $___46__46__47_loader_47_LoaderCompiler_46_js__ && $___46__46__47_loader_47_LoaderCompiler_46_js__.__esModule && $___46__46__47_loader_47_LoaderCompiler_46_js__ || {default: $___46__46__47_loader_47_LoaderCompiler_46_js__}).LoaderCompiler;
var NodeLoaderCompiler = function($__super) {
  function NodeLoaderCompiler() {
    $__superConstructor(NodeLoaderCompiler).call(this);
    this.sourceMapsInMemory_ = false;
  }
  return ($__createClass)(NodeLoaderCompiler, {
    evaluateCodeUnit: function(codeUnit) {
      var runInThisContext = require('vm').runInThisContext;
      var semver = require('semver');
      var content = codeUnit.metadata.transcoded;
      var filename = codeUnit.address || codeUnit.normalizedName;
      if (codeUnit.metadata.traceurOptions.sourceMaps === 'memory') {
        this.enableMemorySourceMaps_();
      }
      var options;
      if (semver.gte(process.version, '0.12.0')) {
        options = {filename: filename};
      } else {
        options = filename;
      }
      var result = runInThisContext(content, options);
      codeUnit.metadata.transformedTree = null;
      return result;
    },
    enableMemorySourceMaps_: function() {
      if (this.sourceMapsInMemory_) {
        return;
      }
      require('source-map-support').install({retrieveSourceMap: function(url) {
          try {
            var map = System.getSourceMap(url);
            if (map) {
              return {
                url: url,
                map: map
              };
            }
          } catch (ex) {
            console.error('retrieveSourceMap FAILED ', ex);
          }
        }});
      this.sourceMapsInMemory_ = true;
    }
  }, {}, $__super);
}(LoaderCompiler);
Object.defineProperties(module.exports, {
  NodeLoaderCompiler: {
    get: function() {
      return NodeLoaderCompiler;
    },
    enumerable: true
  },
  __esModule: {value: true}
});
